src/MMCPServer.cpp

```diff
@@ -0,0 +1,810 @@
+}
+
+/**
+ * Script command, Stop our server from listening
+ */
+QPair<bool, QString> MMCPServer::stopServer()
+{
+    if (isListening()) {
+        close();
+        return QPair<bool, QString>(true, qsl("command successful"));
+    }
+
+    return QPair<bool, QString>(false, qsl("unable to stop server, it is not listening"));
+}
+
+
+/**
+ * Allow a client to snoop you (the user).
+ */
+QPair<bool, QString> MMCPServer::allowSnoop(const QVariant& target)
+{
+    MMCPClient* client = clientByNameOrId(target);
+
+    if (client != nullptr) {
+        if (client->canSnoop()) {
+            client->setCanSnoop(false);
+            client->setSnooped(false);
+            client->sendMessage(QString("<CHAT> You are no longer allowed to snoop %1.").arg(m_chatName));
+
+            const QString infoMsg = tr("[ CHAT ]  - %1 is no longer allowed to snoop you.").arg(client->chatName());
+            mpHost->postMessage(infoMsg);
+
+        } else {
+            client->setCanSnoop(true);
+            client->sendMessage(QString("<CHAT> You are now allowed to snoop %1.").arg(m_chatName));
+
+            const QString infoMsg = tr("[ CHAT ]  - %1 can now snoop you.").arg(client->chatName());
+            mpHost->postMessage(infoMsg);
+        }
+
+        return QPair<bool, QString>(true, qsl("command successful"));
+    }
+
+    return QPair<bool, QString>(false, qsl("no client by that name"));
+}
+
+/**
+ * Send a request to snoop someone, or stop snooping them if snooped
+ */
+QPair<bool, QString> MMCPServer::snoop(const QVariant& target)
+{
+    MMCPClient* client = clientByNameOrId(target);
+
+    if (client != NULL) {
+        if (client->isSnooped()) {
+            client->setSnooped(false);
+        } else {
+            client->snoop();
+        }
+
+        return QPair<bool, QString>(true, qsl("command successful"));
+    }
+
+    return QPair<bool, QString>(false, qsl("no client by that name"));
+}
+
+/**
+ * Unchat someone by name or id
+ */
+QPair<bool, QString> MMCPServer::unChat(const QVariant& target)
+{
+    MMCPClient* client = clientByNameOrId(target);
+
+    if (client != NULL) {
+        client->disconnect();
+        return QPair<bool, QString>(true, qsl("command successful"));
+    }
+
+    return QPair<bool, QString>(false, qsl("no client by that name"));
+}
+
+
+void MMCPServer::addConnectedClient(MMCPClient* client)
+{
+    clients.append(client);
+    client->setId(clients.indexOf(client));
+    emit clientConnected(client);
+}
+
+void MMCPServer::slotClientDisconnected(MMCPClient* client)
+{
+    clients.removeOne(client);
+    emit clientDisconnected(client);
+}
+
+
+/**
+ * Send a message to the terminal pane.
+ * This converts a normal QString into an AttributedString, prepends a carriage return,
+ * and encapsulates the string in the default chat color before sending the result inward.
+ */
+void MMCPServer::clientMessage(const QString& message)
+{
+    QString trimmed = message;
```

Comment:
Have you considered the simpler:
```suggestion
    QString trimmed = message.trimmed();
```
A quick check with a TinTin++ script that sends a single line message every 10s was arriving with a line-feed at either end - and given the pair that are also added when the text is recoloured this is producing a lot more empty lines than I would expect...

Reply:
I need to examine the whole line spacing thing more closely, I've been fighting it for several versions... The proper behavior should be to display a single empty line above and below a chat message.

Reply:
The `[ XXXX ] - message.` type message display system does work to ensure that the first line does start a new line, and IIRC also that it finishes a line. However otherwise it can be tricky to ensure that a piece of text one thing generates does not get mixed up with other bits from other sources; also, extra blank lines can also be an issue for some users - they don't want the (possibly valuable) on-screen display space wasted with bogus lines. TBH I would urge you not to put in empty lines around stuff... :pray: